bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,348 @@
+
+    /**
+     * A {@link BoundedSource} for a Bigtable {@link Table} with a start/stop key range, along with
+     * a potential filter via a {@link Scan}.
+     */
+    private static class SourceWithKeys extends BoundedSource<Result> {
+      private static final long serialVersionUID = 2561066007121040429L;
+
+      private final CloudBigtableScanConfiguration configuration;
+      private final byte[] startKey;
+      private final byte[] stopKey;
+
+      public SourceWithKeys(
+          CloudBigtableScanConfiguration configuration, byte[] startKey, byte[] stopKey) {
+        this.configuration = configuration;
+        this.startKey = startKey;
+        this.stopKey = stopKey;
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        // TODO(sduskis): Hook up the right size for the source.  The Cloud Bigtable API has the
+        // capability to get the size, but it's not currently exposed via HBase. (issue #400)
+        return 0;
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return true;
+      }
+
+      @Override
+      public Coder<Result> getDefaultOutputCoder() {
+        return new HBaseResultCoder();
+      }
+
+      /**
+       * This does not actually split.  It just returns this.  This ought to be fixed.
+       */
+      @Override
+      public List<? extends BoundedSource<Result>> splitIntoBundles(long desiredBundleSizeBytes,
+          PipelineOptions options) throws Exception {
+        return Collections.singletonList(this);
+      }
+
+      @Override
+      public void validate() {
+        checkNotNullOrEmpty(configuration.getProjectId(), "project");
+        checkNotNullOrEmpty(configuration.getZone(), "zone");
+        checkNotNullOrEmpty(configuration.getCluster(), "cluster");
+        checkNotNullOrEmpty(configuration.getTable(), "table");
+      }
+
+      @Override
+      public BoundedSource.BoundedReader<Result> createReader(PipelineOptions options)
+          throws IOException {
+        return new CloudBigtableIO.Source.Reader(this, configuration, startKey, stopKey);
+      }
+    }
+
+    /**
+     * Reads rows for a specific {@link Table}, potentially filted by a Scan and/or a start/stop key
+     * range.
+     */
+    private static class Reader extends BoundedReader<Result> {
+      private final BoundedSource<Result> source;
+      private final Scan scan;
+      private final CloudBigtableScanConfiguration config;
+
+      private Connection connection;
+      private ResultScanner scanner;
+      private Table table;
+      private Result current;
+
+      public Reader(BoundedSource<Result> source, CloudBigtableScanConfiguration config,
+          byte[] startRow, byte[] endRow) throws IOException {
+        super();
+        this.source = source;
+        this.config = config;
+        if (startRow != null || endRow != null) {
```

Comment:
nit: Maybe it will be clearer if this is split into two constructors instead of using nulls? One for a reader bounded with start and end rows, and one for an.. unbounded reader?